### PR TITLE
1150 - add array shuffling on the server prefetch

### LINKED ISSUE #1150: [Bug] Hydration error on the Campaigns Page
**Describe the bug**
image

**To Reproduce**

1. Go to /campaigns on the local environment
2. See error

### PR DESCRIPTION
**Pull request by dimitur2204**
## Motivation and context
Fixes #1150.

Upgraded `react-query` to the newest version, which includes a fix to properly hydrate after React 18.

Changes:
* We now **have to** provide an array of queryKeys, instead of a single one as was possible before. 
* Import has changed from `react-query` to `@tanstack/react-query` 

## Other fixes and notes
Please note the change from `queryFn` to `queryFnFactory` I did it, because of Typing issues, but the `queryFnFactory` has some additonal logic, I hope that this does not slow down the queries or introduce problems.

Fixed LocalizationProvider to now work properly and the translations issues on the profile page are resolved [1150 - fix localization provider locale to new api](https://github.com/podkrepi-bg/frontend/pull/1152/commits/cd81ae5215694c72aa441d0306b0fde0dde54f94)
## Testing
Tried the list and it now seems to be both shuffled and without an error.

### CODE REVIEW

**Comment by kachar** [comment]
@dimitur2204 This probably also should fix #1111

File: src/common/util/shuffle.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
...
-  for (let i = array.length - 1; i > 0; i--) {
+export function shuffleArray<T>(array: T[]) {
+  const arrCopy = [...array]
+  for (let i = arrCopy.length - 1; i > 0; i--) {
     const j = Math.floor(Math.random() * (i + 1))
-    ;[array[i], array[j]] = [array[j], array[i]]
+    ;[arrCopy[i], arrCopy[j]] = [arrCopy[j], arrCopy[i]]
   }
+  return arrCopy
 }
```
Comment (kachar) [inline src/common/util/shuffle.ts:8]: @dimitur2204 We use `lodash` in the project so I find it easy to replace this custom random fn with lodash util for example `shuffle`. 

```tsx
_.shuffle([1, 2, 3, 4]);
```

https://lodash.com/docs/4.17.15#shuffle

**Review by igoychev — CHANGES_REQUESTED** [review]
Was quite happy that you found way to fix, however just tested and unfortunately it looks like the shuffling does not work anymore. 

What I see is that campaigns get shuffled, but a quick flicker returns them to always the same order. Tried with 3 tabs and 2 browsers (chrome/firefox) and all have the same order for all campaigns after refreshing.

File: src/common/util/shuffle.ts (hunk 1 above)
Comment (dimitur2204) [reply src/common/util/shuffle.ts:8]: Will do that!

**Comment by dimitur2204** [comment]
The issue was fixed, but something else occurred now after the update. I noticed we have a couple of pages that do this check. This way for a second between the DOM is loaded we can see a `<NotFoundPage/>`. @kachar I feel like we shouldn't be returning a not found page like that. What do you say?
image

**Comment by kachar** [comment]
> The issue is fixed, but something else occurred now after the update. I noticed we have a couple of pages that do this check. This way for a second until the DOM is loaded we can see a `<NotFoundPage/>`. @kachar I feel like we shouldn't be returning a not found page like that. What do you say? image

I think we need to use `getServerSideProps` which returns `{ notFound: true }` if the campaign is not available in the API.

**Review by valkirilov — APPROVED** [review]
Good job! It seems that the hydration issue is fixed after some tests on my local, I confirm it. Although I still don't feel confident enough and I'm still exploring the website and "learning" the different features which are available.

PS: I can confirm a performance improvement as well, the site is working way more faster than before. Kudos :clap:

**Review by igoychev — APPROVED** [review]

**Comment by igoychev** [comment]
Amazing job Nizamov!

**Review by kachar — APPROVED** [review]
Kudos for the enhancement @dimitur2204 

Works great now!

**Pull request merged**